[17.0][ADD] website_sale_cart_add_product_xlsx_csv

Tisho99 opened the pull request:
This module adds a button in the website cart that let users import a xls or csv file that contains the products and quantities to add to the cart

T-8854

Tisho99 committed:
[ADD] website_sale_cart_add_product_xlsx_csv

luis-ron reviewed (approved):
Functional review: LGTM 👍🏻

ValentinVinagre commented on requirements.txt:
```diff
@@ -0,0 +1,3 @@
+# generated from manifests external_dependencies
+pandas
```
Why do we need the pandas library? It's a very heavy library.
---> https://docs.python.org/3/library/csv.html

Tisho99 replied:
Its used to parse the xls and the csv. Its the library i knew, and what was first recomended in the searches i made for xls and csv processing. Is there a better alternative? Why?

ValentinVinagre replied:
It's a very heavy library, as I mentioned before. I've given you an alternative for CSV; Odoo already uses libraries for this; check which ones it uses.

ValentinVinagre commented on website_sale_cart_add_product_xlsx_csv/data/import_file_example.xml:
```diff
@@ -0,0 +1,16 @@
+        />
+        <field
+            name="url"
+        >website_sale_cart_add_product_xlsx_csv/static/xls/import_example.xls</field>
```
why extensión xls and not xlsx, does it support both extensions?

Tisho99 replied:
I'll change it

ValentinVinagre commented on website_sale_cart_add_product_xlsx_csv/models/website.py:
```diff
@@ -0,0 +1,12 @@
+    _inherit = "website"
+
+    cart_import_button = fields.Boolean(
+        string="Is the cart import button enabled?", default=True
```
default=False

Tisho99 replied:
Ok, I'll change it

ValentinVinagre commented on website_sale_cart_add_product_xlsx_csv/controllers/main.py:
```diff
@@ -0,0 +1,131 @@
+
+
+class WebsiteSaleAddProductXlsCsv(WebsiteSale):
+    def can_be_imported(self, product, qty, sale):
```
It is indicated which products?

Tisho99 replied:
It has not been indicated, neither in the readme nor in the docstring. However, i think it can be overexplaining because the code is clear. What do you think i have to indicate?

ValentinVinagre replied:
If I upload a file with 300 lines and 20 lines fail, how does the user know which ones are failing and why?

Tisho99 replied:
I had't understood your question, let me explain it better:

A error msg is given to the user with the internal_code of the products that failed. If it is not enought for large files, i can add the line numbers that are failing

ValentinVinagre replied:
change it

ValentinVinagre commented on website_sale_cart_add_product_xlsx_csv/controllers/main.py:
```diff
@@ -0,0 +1,131 @@
+                    "the columns should be 'default_code' and 'product_uom_qty'"
+                )
+
+        if import_status != "error":
```
why other if ? Doesn't it make more sense for it to be an else?

Tisho99 replied:
It must be an if because the import_status could have changed inside the condition above to a error

ValentinVinagre commented on website_sale_cart_add_product_xlsx_csv/controllers/main.py:
```diff
@@ -0,0 +1,131 @@
+        import_msg = ""
+        failed_products = []
+
+        if file.filename.endswith(".csv"):
```
The file size should be checked. Could you set a system parameter to indicate the size in MB? The default is 2MB, but what do you think?

Tisho99 replied:
Ok. I'll do it

Tisho99 commented:
Hello @ValentinVinagre 

I have made the requested changes. Its ready for review.

However, this might need a little more refactor because the new functionalities and libraries have enlarged too much the main functions. I'd like to know your opinion first

Could you review please?

luis-ron reviewed (approved):
Functional review: LGTM 👍🏻

Anxo82 commented:
Technical review: LGTM :+1:

Tisho99 commented:
Hello @PedroBaeza,

Could you please review this PR when you have time?

It adds a button in the website cart that allows users to import an XLS or CSV file with products and quantities to add them to the cart.

Thank you!

pedrobaeza commented:
@ValentinVinagre should re-review his requested changes. About the module itself, shouldn't be this something to be added in the portal (next to the re-order option) instead of the cart? If not, you need to have something in the cart to access the feature.

ValentinVinagre commented:
Hi @pedrobaeza , initially the shopping cart button is visible even when it's empty.
[Grabación de pantalla desde 2026-03-09 15-39-03.webm](https://github.com/user-attachments/assets/52e86a9d-6bfa-47db-9865-a943a65e9ea7)

pedrobaeza commented:
OK, I think that's configurable be website, so if hidden, not able to get it, but if you consider it correct, OK. Please remove your requested changes if everything is OK.

ValentinVinagre commented:
@Tisho99 could u add to roadmap "If the shopping cart button is hidden without anything in it, the functionality cannot be used until something is added to the cart." ?

ValentinVinagre reviewed (approved):
LGTM 👍🏻

Tisho99 commented:
Hello @pedrobaeza,

We have addressed the suggestions you made on this PR,

Whenever you have a moment, could you please take another look?

Thank you very much in advance!

pedrobaeza commented:
/ocabot merge nobump
